src/unit.h

```diff
@@ -6,25 +6,165 @@
+// macro
+
+// Convert angstroms to SI meters
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(datasetunits::wavelength_base_unit,
+                                     boost::units::si::length, double, 1.0e-10);
+// Convert meV to SI Joule
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(
+    datasetunits::energy_base_unit, boost::units::si::energy, double,
+    1.0e-3 * boost::units::si::constants::codata::e.value().value());
+// Convert tof microseconds to SI seconds
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(datasetunits::tof_base_unit,
+                                     boost::units::si::time, double, 1.0e-6);
+
+template <>
+struct boost::units::base_unit_info<datasetunits::wavelength_base_unit> {
+  static std::string name() { return "angstroms"; }
+  static std::string symbol() { return "AA"; }
+};
+
+template <>
+struct boost::units::base_unit_info<datasetunits::counts_base_unit> {
+  static std::string name() { return "counts"; }
+};
+
+template <>
+struct boost::units::base_unit_info<datasetunits::energy_base_unit> {
+  static std::string name() { return "milli-electronvolt"; }
+  static std::string symbol() { return "meV"; }
+};
+
+template <> struct boost::units::base_unit_info<datasetunits::tof_base_unit> {
+  static std::string name() { return "microseconds"; }
+  static std::string symbol() { return "us"; }
+};
+
 class Unit {
 public:
+  // Define a std::variant which will hold the set of allowed units.
+  // Any unit that does not exist in the variant will either fail to compile or
+  // throw a std::runtime_error during operations such as multiplication or
+  // division.
+  // 
+  // TODO: maybe it is possible to create a helper that will automatically
+  // generate the squares for variance?
+  // The following was attempted but did not succeed:
+  //  template <class... Ts> struct unit_and_variance {
+  //  using type =
+  //    std::variant<Ts..., decltype(std::declval<Ts>()*std::declval<Ts>())...>;
+  //  };
+  typedef std::variant<
+      // Dimensionless [ ]
+      boost::units::si::dimensionless,
+      // Length [m]
+      boost::units::si::length,
+      // Area [m^2]
+      boost::units::si::area,
+      // Time [s]
+      boost::units::si::time,
+      // Mass [kg]
+      boost::units::si::mass,
+      // Counts [counts]
+      datasetunits::counts,
+      // InverseLength [m^-1]
+      decltype(std::declval<boost::units::si::dimensionless>() /
+               std::declval<boost::units::si::length>()),
+      // Wavelength [Angstroms]
+      datasetunits::wavelength,
+      // Energy [meV]
+      datasetunits::energy,
+      // Time of flight [microseconds]
+      datasetunits::tof,
+      // 1/time [s^-1]
+      decltype(std::declval<boost::units::si::dimensionless>() /
+               std::declval<boost::units::si::time>()),
+      // Area variance
+      decltype(std::declval<boost::units::si::area>() *
+               std::declval<boost::units::si::area>()),
```

Comment:
Would this whole thing get more readable/maintainable if you define helper units somewhere, so when listing all units you can simply write `decltype(m*m), decltype(m/s)`, etc.?